src/wasm/kotlin/wasm/arcs/Particle.kt

```diff
@@ -0,0 +1,237 @@
+package arcs
+
+import kotlin.collections.set
+
+typealias URL = String
+
+abstract class Entity<T> {
+    var internalId = ""
+    abstract fun decodeEntity(encoded: String): T?
+    abstract fun encodeEntity(): String
+}
+
+/**
+ * Base class for all Wasm Particles.
+ */
+abstract class Particle : WasmObject() {
+    private val handles: MutableMap<String, Handle> = mutableMapOf()
+    private val toSync: MutableSet<Handle> = mutableSetOf()
+    private val eventHandlers: MutableMap<String, () -> Unit> = mutableMapOf()
+
+    fun registerHandle(name: String, handle: Handle) {
+        handle.name = name
+        handles[name] = handle
+        log("Registering $name")
+    }
+
+    fun eventHandler(name: String, handler: () -> Unit) {
```

Comment:
This is a good idea; I'll look at doing something similar in the C++ code (I had a low priority task to do "something better" for the event handling).